File: MadridTransporte.Tests/TimesTests.cs
using System.Net;
using System.Text.Json;
using MadridTransporte.Api.Dtos;
using MadridTransporte.Tests.Fixtures;
using Shouldly;

namespace MadridTransporte.Tests;

public class TimesTests
{
    private const string BusStopCode = "08242";
    private const string TrainStopCode = "34";
    private const string MetroStopCode = "235";
    private const string EmtStopCode = "73";
    private const string TramStopCode = "58";

    [Test]
    [Arguments("/stops/bus/" + BusStopCode + "/times")]
    [Arguments("/stops/train/" + TrainStopCode + "/times")]
    [Arguments("/stops/metro/" + MetroStopCode + "/times")]
    [Arguments("/stops/emt/" + EmtStopCode + "/times")]
    [Arguments("/stops/tram/" + TramStopCode + "/times")]
    public async Task Should_Get_Stop_Times(string url)
    {
        await PostgresFixture.EnsureInitialized();

        var response = await PostgresFixture.Client.GetAsync(url);
        var responseBody = await response.Content.ReadAsStringAsync();
        response.StatusCode.ShouldBe(HttpStatusCode.OK, responseBody);

        var body = JsonSerializer.Deserialize<StopTimesDto>(responseBody, PostgresFixture.JsonOptions);
        body.ShouldNotBeNull();
        body.CodMode.ShouldBeGreaterThan(0);
        body.StopName.ShouldNotBeNullOrEmpty();
        body.SimpleStopCode.ShouldNotBeNullOrEmpty();
        body.Coordinates.ShouldNotBeNull();
        body.Arrives.ShouldNotBeNull();

        foreach (var arrive in body.Arrives!)
        {
            arrive.CodMode.ShouldBeGreaterThan(0);
            arrive.Line.ShouldNotBeNullOrEmpty();
            arrive.Destination.ShouldNotBeNullOrEmpty();
            arrive.EstimatedArrives.ShouldNotBeEmpty();
        }

        foreach (var incident in body.Incidents)
        {
            incident.Title.ShouldNotBeNullOrEmpty();
            incident.Description.ShouldNotBeNullOrEmpty();
        }
    }

    [Test]
    [Arguments("/stops/bus/asdasd/times")]
    [Arguments("/stops/train/asdasd/times")]
    [Arguments("/stops/metro/asdasd/times")]
    [Arguments("/stops/emt/asdasd/times")]
    [Arguments("/stops/tram/asdasd/times")]
    public async Task Should_Not_Get_Stop_Times(string url)
    {
        await PostgresFixture.EnsureInitialized();

        var response = await PostgresFixture.Client.GetAsync(url);
        response.StatusCode.ShouldBe(HttpStatusCode.NotFound);
    }
}

File: tests/src/test/kotlin/stops/TimesTests.kt
package stops

import arrow.core.getOrElse
import arrow.core.raise.either
import common.extensions.getOrThrow
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeInstanceOf
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.EnumSource
import simpleJson.*
import utils.testApplicationBusTracker

const val busStopCode = "08242"
const val trainStopCode = "34"
const val metroStopCode = "235"
const val emtStopCode = "73"
const val emtStopCode2 = "77"
const val tramStopCode = "58"

enum class Times(val url: String) {
    BUS("/stops/bus/$busStopCode/times"),
    TRAIN("/stops/train/$trainStopCode/times"),
    METRO("/stops/metro/$metroStopCode/times"),
    EMT("/stops/emt/$emtStopCode/times"),
    EMT2("/stops/emt/$emtStopCode2/times"),
    TRAM("/stops/tram/$tramStopCode/times")
}

enum class TimesNotFound(val url: String) {
    BUS("/stops/bus/asdasd/times"),
    TRAIN("/stops/train/asdasd/times"),
    METRO("/stops/metro/asdasd/times"),
    EMT("/stops/emt/asdasd/times"),
    TRAM("/stops/tram/asdasd/times")
}

class TimesTests {
    @ParameterizedTest
    @EnumSource(Times::class)
    fun `should get stop times`(code: Times) = testApplicationBusTracker {
        val response = client.get(code.url)
        val body = response.bodyAsText().deserialized().getOrElse { throw it }

        response.status.value.shouldBeEqualTo(200)
        body.shouldBeInstanceOf<JsonObject>()
        either {
            body["codMode"].asInt().bind()
            body["stopName"].asString().bind()
            body["simpleStopCode"].asString().bind()
            body["stopCode"].asString().bind()
            body["coordinates"]["latitude"].asDouble().bind()
            body["coordinates"]["longitude"].asDouble().bind()
            body["arrives"].asArray().bind().forEach { arrive ->
                arrive["codMode"].asInt().bind()
                arrive["line"].asString().bind()
                arrive["anden"].asInt().getOrNull()
                arrive["destination"].asString().bind()
                arrive["estimatedArrives"].asArray().bind().map { it.asLong().bind() }.first()
            }
            body["incidents"].asArray().bind().forEach {
                it["title"].asString().bind()
                it["description"].asString().bind()
                it["from"].asString().bind()
                it["to"].asString().bind()
                it["cause"].asString().bind()
                it["effect"].asString().bind()
                it["url"].asString().bind()
            }
        }.getOrThrow()
    }

    @ParameterizedTest
    @EnumSource(TimesNotFound::class)
    fun `should not get stop times`(code: TimesNotFound) = testApplicationBusTracker {
        val response = client.get(code.url)
        response.status shouldBeEqualTo HttpStatusCode.NotFound
    }
}